New Podcast Episode: Oxide with Steve Klabnik (#51)
corrode/corrode.github.io: content/podcast/s03e03-oxide/index.md

+++
title = "Oxide"
date = 2024-11-14
template = "episode.html"
draft = false
aliases = ["/p/s03e03"]
[extra]
guest = "Steve Klabnik"
role = "Author and Software Engineer"
season = "03"
episode = "03"
series = "Podcast"
+++

> 

Reviewer: ```suggestion

<div><script id="letscast-player-97aa1056" src="https://letscast.fm/podcasts/rust-in-production-82281512/episodes/oxide-s-steve-klabnik/player.js?size=s"></script></div>

```

<!-- more -->

## Show Notes

### About Oxide Computer Company

Founded by industry giants Bryan Cantrill, Jessie Frazelle, and Steve Tuck, Oxide Computer Company is a beloved name in the Rust community. They took on the daunting task of rethinking how servers are built -- starting all the way from the hardware and boot process (and no, there is no BIOS).
Their 'On The Metal' podcast is a treasure trove of systems programming stories and proudly served as a role model for 'Rust in Production.'

### About Steve Klabnik

In the Rust community, Steve does not need any introduction. He is a prolific writer, speaker, and software engineer who has contributed to the Rust ecosystem in many ways -- including writing the first version of the official Rust book. If you sent a tweet about Rust in the early days, chances are Steve was the one who replied. Previously, he worked at Mozilla and was a member of the Rust and Ruby core teams. 

### Links From The Episode (In Chronological Order)

- [The Rust Programming Language](https://doc.rust-lang.org/book/) ([No Starch Press version](https://nostarch.com/Rust)) - The official Rust book

> - Steve's Rust 1.0 [FOSDEM](https://archive.fosdem.org/2015/schedule/event/the_story_of_rust/) / [ACM](https://www.youtube.com/watch?v=79PSagCD_AY) talk - Early history of Rust

Reviewer: ```suggestion
- [The Story of Rust - FOSDEM](https://archive.fosdem.org/2015/schedule/event/the_story_of_rust/) / [The History of Rust - ACM](https://www.youtube.com/watch?v=79PSagCD_AY) - Early history of Rust
```

- [The Soul of a New Machine](https://en.wikipedia.org/wiki/The_Soul_of_a_New_Machine) by Tracy Kidder - Classic book on computer engineering
- [I have come to bury the BIOS](https://www.osfc.io/2022/talks/i-have-come-to-bury-the-bios-not-to-open-it-the-need-for-holistic-systems/) - Bryan's talk on firmware
- [Beowulf cluster](https://en.wikipedia.org/wiki/Beowulf_cluster) - Early parallel computing architecture
- [Bryan's blog post on Rust](https://bcantrill.dtrace.org/2018/09/18/falling-in-love-with-rust/) - Journey of a systems programmer to Rust
- [JavaOS](https://en.wikipedia.org/wiki/JavaOS) - Operating system written in Java
- [D Programming Language](https://dlang.org/) - Systems programming language
- [Garbage Collection in early Rust](https://pcwalton.github.io/_posts/2013-06-02-removing-garbage-collection-from-the-rust-language.html) - Historical Rust development
- [Removing green threads RFC](https://github.com/rust-lang/rfcs/blob/master/text/0230-remove-runtime.md) - Major change in Rust's concurrency model
- [Hubris](https://github.com/oxidecomputer/hubris) - Oxide's embedded operating system

> - [Tock OS](https://tockos.org/) - Embedded operating system in Rust

Reviewer: ```suggestion
- [Tock OS](https://tockos.org/) - Embedded operating system in Rust predating Hubris
```

- [cargo-xtask](https://github.com/matklad/cargo-xtask) - Build automation for Rust projects

> - [Hubris Build Documentation](https://github.com/oxidecomputer/hubris#build) - Building Hubris

Reviewer: ```suggestion
- [Hubris Build Documentation](https://github.com/oxidecomputer/hubris#build) - Building Hubris using `cargo xtask`
```

- [Buck Build System](https://buck.build/)
- [Buildomat](https://github.com/oxidecomputer/buildomat)
- [Omicron](https://github.com/oxidecomputer/omicron)
- [illumos](https://illumos.org/)
- [bhyve](https://bhyve.org/)
- [About Self-hosted Runners](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners)
- [Async Drop Initiative](https://rust-lang.github.io/async-fundamentals-initiative/roadmap/async_drop.html)

> - [Rust Playground Example for async Rust syntax](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=async+fn+f()+%7B%0A++++await+f()%0A%7D)

Reviewer: ```suggestion
- [Rust Playground Example for helpful error when using prefix `await` operator](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=async+fn+f()+%7B%0A++++await+f()%0A%7D)
```

> - [bhyve](https://bhyve.org/) - Hypervisor

Reviewer: ```suggestion
- [bhyve](https://bhyve.org/) - BSD VM hypervisor
```

- [Rust Book - Modules](https://doc.rust-lang.org/book/ch07-02-defining-modules-to-control-scope-and-privacy.html)
- [OpenAPI Specification](https://swagger.io/specification/)
- [Dropshot](https://github.com/oxidecomputer/dropshot/)
- [Axum](https://github.com/tokio-rs/axum)
- [Oxide Web Console](https://github.com/oxidecomputer/console)
- [Oxide Web Console - Preview](https://console-preview.oxide.computer/projects)
- [Oxide RFD 1](https://rfd.shared.oxide.computer/rfd/0001)
- [Rust RFCs](https://rust-lang.github.io/rfcs/)
- [IETF RFCs](https://en.wikipedia.org/wiki/Request_for_Comments)
- [Zig](https://ziglang.org/)
- [TigerBeetle](https://tigerbeetle.com/)
- [Bun](https://bun.sh/)
- [CockroachDB](https://www.cockroachlabs.com/)
- [Oxide and Friends: Wither CockroachDB?](https://oxide.computer/podcasts/oxide-and-friends/2052742)
- [Mozilla Public License](https://choosealicense.com/licenses/mpl-2.0/)

> - [Asahi Linux Rust Drivers](https://asahilinux.org/)

Reviewer: ```suggestion
- [Asahi Linux uses Rust to write drivers](https://asahilinux.org/)
```

> - [Dropshot](https://github.com/oxidecomputer/dropshot/) - Oxide's HTTP framework

Reviewer: ```suggestion
- [Dropshot](https://github.com/oxidecomputer/dropshot/) - Oxide's OpenAPI server framework
```

> - [Oxide Console Preview](https://console-preview.oxide.computer/projects) - Demo of Oxide Console

Reviewer: ```suggestion
- [Oxide Console Preview](https://console-preview.oxide.computer/projects) - Demo of Oxide Console using a mocked backend
```

> - [TigerBeetle](https://tigerbeetle.com/) - Financial accounting database

Reviewer: ```suggestion
- [TigerBeetle](https://tigerbeetle.com/) - Financial accounting database written in Zig
```

> - [Bun](https://bun.sh/) - JavaScript runtime

Reviewer: ```suggestion
- [Bun](https://bun.sh/) - JavaScript toolkit written in Zig
```

> - [CockroachDB](https://www.cockroachlabs.com/) - Distributed SQL database

Reviewer: ```suggestion
- [CockroachDB](https://www.cockroachlabs.com/) - Distributed SQL database used in Oxide's backplane
```

> - [Oxide and Friends: Wither CockroachDB?](https://oxide.computer/podcasts/oxide-and-friends/2052742) - Podcast episode

Reviewer: ```suggestion
- [Oxide and Friends: Wither CockroachDB?](https://oxide.computer/podcasts/oxide-and-friends/2052742) - Discussing Cockroach's switch away from BUSL
```

> - [Mozilla Public License](https://choosealicense.com/licenses/mpl-2.0/) - Software license

Reviewer: ```suggestion
- [Mozilla Public License](https://choosealicense.com/licenses/mpl-2.0/) - Oxide's default software license
```

- [Buck2](https://buck2.build/) - Meta's build system
- [Jujutsu (jj)](https://martinvonz.github.io/jj/) - Git replacement
- [Steve's Jujutsu Tutorial](https://steveklabnik.github.io/jujutsu-tutorial/introduction/introduction.html) - Guide to jj

> - [Steve's blog post on branch naming](https://steveklabnik.com/writing/against-names/)

Reviewer: ```suggestion
- [Steve's blog post on not naming branches](https://steveklabnik.com/writing/against-names/)
```
